server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1252,8 +1254,41 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                 }
             });
 
+        final StepListener<RepositoryData> filterRepositoryDataStep = new StepListener<>();
+
         // Step 2: Write new index-N blob to repository and update index.latest
         setPendingStep.whenComplete(newGen -> threadPool().executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(listener, l -> {
+            // BwC logic: Load snapshot version information if any snapshot is missing a version in RepositoryData so that the new
+            // RepositoryData contains a version for every snapshot
+            final List<SnapshotId> snapshotIdsWithoutVersion = repositoryData.getSnapshotIds().stream().filter(
+                snapshotId -> repositoryData.getVersion(snapshotId) == null).collect(Collectors.toList());
+            if (snapshotIdsWithoutVersion.isEmpty() == false) {
+                final Map<SnapshotId, Version> updatedVersionMap = new ConcurrentHashMap<>();
+                final GroupedActionListener<Void> loadAllVersionsListener = new GroupedActionListener<>(
+                    ActionListener.runAfter(
+                        new ActionListener<>() {
+                            @Override
+                            public void onResponse(Collection<Void> voids) {
+                                logger.info("Successfully loaded all snapshot's version information for {} from snapshot metadata",
```

Comment:
`info` IMO since it's a one time thing and it would be nice to have some marker of the "upgrade" in the logs.

Reply:
Given that `snapshotIdsWithoutVersion` can be very long. I was wondering if we should just display the size of `snapshotIdsWithoutVersion` here in the info logging.

Reply:
I think it would be nice to have the concrete ids of the snapshots to help debug situations where users might be writing to the repo from various ES versions (that's my main motivation to have this). I figured the list isn't going to be so long that it would create real issues due to the log line length so worst case it's a bit of an annoyance right?

Reply:
> I think it would be nice to have the concrete ids of the snapshots to help debug situations where users might be writing to the repo from various ES versions (that's my main motivation to have this)

Why do the exact names and versions matter in that case? Isn't it sufficient to know that some snapshots got version-tagged that previously were not?

Reply:
Versions I don't care about much and the exact names aren't important either, but it would be nice to see if the list of snapshots changed if this gets logged repeatedly.
Especially on Cloud just having just the size which might be effectively constant over time seems like it could create an annoying situation when debugging :)

Reply:
Perhaps use `firstListElementsToCommaDelimitedString` then (see AllocationService)?

Reply:
Ah nice thanks => reused in  https://github.com/elastic/elasticsearch/pull/50930/commits/02e8c4ee3f0ef38a2fc059f5583c29e8f5a133a3 (can probably move that method to a more appropriate place in a follow up)